x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/new_terms/utils.ts

```diff
@@ -46,3 +50,94 @@ export const validateHistoryWindowStart = ({
     );
   }
 };
+
+/**
+ * Takes a list of buckets and creates value from them to be used in 'include' clause of terms aggregation.
+ * For a single new terms field, value equals to bucket name
+ * For multiple new terms fields and buckets, value equals to concatenated base64 encoded bucket names
+ * @returns for buckets('host-0', 'test'), resulted value equals to: 'aG9zdC0w_dGVzdA=='
+ */
+export const transformBucketsToValues = (
+  newTermsFields: string[],
+  buckets: estypes.AggregationsCompositeBucket[]
+): Array<string | number> => {
+  // if new terms include only one field we don't use runtime mappings and don't stich fields buckets together
+  if (newTermsFields.length === 1) {
+    return buckets
+      .map((bucket) => Object.values(bucket.key)[0])
+      .filter((value): value is string | number => value != null);
+  }
+
+  return buckets.map((bucket) =>
+    Object.values(bucket.key)
+      .filter((value): value is string | number => value != null)
+      .map((value) =>
+        Buffer.from(typeof value !== 'string' ? value.toString() : value).toString('base64')
+      )
+      .join(DELIMITER)
+  );
+};
+
+export const getNewTermsRuntimeMappings = (
+  newTermsFields: string[]
+): undefined | { [AGG_FIELD_NAME]: estypes.MappingRuntimeField } => {
+  // if new terms include only one field we don't use runtime mappings and don't stich fields buckets together
+  if (newTermsFields.length === 1) {
+    return undefined;
+  }
+
+  return {
+    [AGG_FIELD_NAME]: {
+      type: 'keyword',
+      script: {
+        params: { fields: newTermsFields },
+        source: `
+          void traverseDocFields(def doc, def fields, def index, def line) {
+            if (index === fields.length) {
+              emit(line);
+            } else {
+              for (field in doc[fields[index]]) {
+                def delimiter = index === 0 ? '' : '${DELIMITER}';
+                def nextLine = line + delimiter + String.valueOf(field).encodeBase64();
+
+                traverseDocFields(doc, fields, index + 1, nextLine);
+              }
+            }
+          }
+
+          traverseDocFields(doc, params['fields'], 0, '');
```

Comment:
This is awesome! Definitely the most complicated runtime field I've seen so far, but well worth it for this feature IMO. 
A couple questions:
1. Does anything break if we select multiple fields that each have many values? Certain fields such as [process.args](https://www.elastic.co/guide/en/ecs/current/ecs-process.html#field-process-args) can be very large arrays of values, so even though we limit it to 3 fields, if each one has 100 values we'd be emitting 1,000,000 values here. Adding a hard cap on the number of emitted values could be useful, if ES doesn't have a circuit breaker built in already.
2. Some integration test coverage to execute this runtime field specifically against test data and verify that it creates the expected values, particularly in edge cases like null values and large arrays, would be beneficial.

Reply:
> Does anything break if we select multiple fields that each have many values? Certain fields such as [process.args](https://www.elastic.co/guide/en/ecs/current/ecs-process.html#field-process-args) can be very large arrays of values, so even though we limit it to 3 fields, if each one has 100 values we'd be emitting 1,000,000 values here. Adding a hard cap on the number of emitted values could be useful, if ES doesn't have a circuit breaker built in already.

I've tried to perform test on large array: there is a limit on 100 emitted fields

```
    {
        "shard": 0,
        "index": "array-high-cardinality-data-fake_hosts-2022-11-04",
        "node": "yZmoJFdNRHe4uovryh5mDw",
        "reason": {
          "type": "script_exception",
          "reason": "runtime error",
          "script_stack": [
            "org.elasticsearch.server@8.6.0-SNAPSHOT/org.elasticsearch.script.AbstractFieldScript.checkMaxSize(AbstractFieldScript.java:125)",
            "org.elasticsearch.server@8.6.0-SNAPSHOT/org.elasticsearch.script.StringFieldScript$Emit.emit(StringFieldScript.java:145)",
            """emit(line);
            } else {
              """,
            "     ^---- HERE"
          ],
          "script": " ...",
          "lang": "painless",
          "position": {
            "offset": 140,
            "start": 135,
            "end": 182
          },
          "caused_by": {
            "type": "illegal_argument_exception",
            "reason": "Runtime field [new_terms_values] is emitting [101] values while the maximum number of values allowed is [100]"
          }
        }
      },
```
Which is, doesn't look to many indeed, but would prevent overuse of resources on ES side

Reply:
@marshallmain 
Added following functionality
- ensured that runtime field script does not emit more than 100 fields
- added tests for preview that cover arrays with large number of items

TODO:
- add tests for runtime script

Reply:
added tests for runtime script https://github.com/elastic/kibana/pull/143943/commits/bc50e91d1fea8b57fa62f733dddd6a08b8274b88